README.md
dovesnap
=================

dovesnap is a docker network provider, that works with FAUCET and OVS. This allows docker networks to make use of FAUCET's features, such as mirroring, ACLs, and Prometheus based monitoring.

Thanks to the folks who wrote the orginal [docker-ovs-plugin](https://github.com/gopher-net/docker-ovs-plugin) which is what this project was forked from.

See also https://docs.faucet.nz for FAUCET documentation, including monitoring documentation (dovesnap will supply FAUCET-based monitoring without needing configuration, when started as below).

### Requirements

* Linux host running a supported version docker (x86 and Pi are supported)
* Optionally: additional physical interfaces to connect other hosts also running dovesnap
* non-netfilter iptables. For Debian/Ubuntu, follow the legacy option at https://wiki.debian.org/iptables (`update-alternatives --set iptables /usr/sbin/iptables-legacy`). This requirement will be addressed in a future version.

### Installing as a systemd service

The `install.sh` and `uninstall.sh` scripts can be used to install and uninstall dovesnap as a systemd managed service. An upgrade can be accomplished by executing a `git pull` within `~dovesnap` as the `dovesnap` user, and restarting the service. Persistent configuration is stored in `~dovesnap/service.env`.

### QuickStart Instructions

These instructions describe the most basic use of dovesnap - creating a docker network with Internet access, where dovesnap provides all the FAUCET infrastructure. See below for more advanced usage.

**1.** Make sure you are using Docker 1.10 or later

**2.** You need to `modprobe openvswitch` on the machine where the Docker Daemon is located. Make sure that while the module is loaded, OVS is not running on the host.

```
$ sudo modprobe openvswitch
```

**3.** Create a directory for FAUCET to store its configuration:

```
$ sudo mkdir /etc/faucet
```

**4.** Start dovesnap.

`$ docker-compose build && docker-compose -f docker-compose.yml -f docker-compose-standalone.yml up -d`

**5.** Now you are ready to create a new network

```
$ docker network create mynet -d dovesnap --internal -o ovs.bridge.mode=nat -o ovs.bridge.dpid=0x1 -o ovs.bridge.controller=tcp:127.0.0.1:6653,tcp:127.0.0.1:6654
```

`-d dovesnap` tells docker to use dovesnap as a network provider.

`--internal` tells docker not to supply an additional network connection to containers on the new network for internet access. This is essential for dovesnap to complete control over connectivity.

`-o ovs.bridge.mode=nat` tells dovesnap to arrange NAT for the new network.

`-o ovs.bridge.dpid=0x1 -o ovs.bridge.controller=tcp:127.0.0.1:6653,tcp:127.0.0.1:6654` tell dovesnap which FAUCET will control this network (you can provide your own FAUCET elsewhere on the network, but in this example we are using a dovesnap-provided FAUCET instance).

**6.** Test it out!

```
$ docker run -d --net=mynet --rm --name=testcon busybox sleep 1h
$ docker exec -t testcon ping 4.2.2.2
```

### Advanced usage

There are several options available when creating a network, and when creating a container on a network, to access FAUCET features.

You can view dovesnap's OVS bridges using `ovs-vsctl` and `ovs-ofctl`, from within the dovesnap OVS container. You can even use `ovs-vsctl` to add other (for example, physical) ports to a dovesnap managed bridge and dovesnap will monitor them for you. However, it's recommended that you use dovesnap's own options (below) where possible.

#### Required options

`ovs.bridge.dpid=<dpid> -o ovs.bridge.controller=tcp:<ip>:<port>`

Every dovesnap network requires a DPID (for OVS and FAUCET), and at least one controller for OVS. dovesnap will provide FAUCET and Gauge processes to do forwarding and monitoring by default - at least one FAUCET is required, and one Gauge if monitoring is desired.

#### Network options

These options are supplied at `docker network create` time.

##### Bridge modes

There are three `ovs.bridge.mode` modes, `flat`, `nat`, and `routed`. The default mode is `flat`.

- `flat` causes dovesnap to provide connectivity only between containers on this docker network - not to other networks (essentially, provide a VLAN - no routing).

- `nat` causes dovesnap to provision a gateway, and NAT, for the docker network.

- `routed` causes dovesnap to provision a gateway, for the docker network. An upstream network may provide NAT if needed.

If NAT is in use, you can specify `-p <outside port>:<inside port>` when starting a container. dovesnap will provision a DNAT rule, via the network's gateway from the outside port to the inside port on that container. This mapping won't show up in `docker ps`, as dovesnap is not using docker-proxy.

You can also specify an input ACL for the gateway's port with `-o ovs.bridge.nat_acl=<acl>`, and a default ACL for container ports with `-o ovs.bridge.default_acl=<acl>`.

##### Preallocated ports

`-o ovs.bridge.preallocate_ports=10`

This requests that N ports be pre-allocated with the default ACL (if any) when the network is created. This makes container startup faster, because FAUCET network has already
been configured.

##### Userspace mode

`-o ovs.bridge.userspace=true`

This requests a user space ("netdev"), rather than kernel space switch from OVS. Certain OVS features such as meters, used to implement rate limiting, will only work on a user space bridge.

##### MAC on OVS local port

`-o ovs.bridge.ovs_local_mac=0e:01:00:00:00:03`

This option sets the MAC address of OVS' "local" port on the switch.

You can set the MAC address on a container, with `docker run --mac-address <mac>` (https://docs.docker.com/engine/reference/run/#network-settings)

##### Adding a physical port/real VLAN

`-o ovs.bridge.add_ports=eno123/8`

Dovesnap will connect `eno123` to the Docker network, and attempt to use OVS OFPort 8 (OVS will select another port number, if for some reason port 8 is already in use). You can specify more ports with commas. The OFPort specification is optional - if not present dovesnap will select the next free port number. If specifying a port, you can also specify a third parameter - the ACL name to be applied to the port.

##### Adding a physical port for coprocessing

`-o ovs.bridge.add_copro_ports=eno123/8`

Dovesnap will connect `eno123` to the Docker network as a FAUCET coprocessor port, and attempt to use OVS OFPort 8 (OVS will select another port number, if for some reason port 8 is already in use). You can specify more ports with commas. The OFPort specification is optional - if not present dovesnap will select the next free port number. If specifying a port, you can also specify a third parameter - the ACL name to be applied to the port.

##### Specifying a specific VLAN to use

`-o ovs.bridge.vlan=100`

This adds the VLAN tag of 100 for the Docker network. The default is 100.

#### Specifying an VLAN output ACL to use

`-o ovs.bridge.vlan_out_acl=allowall`

This adds the output ACL `allowall` to the VLAN used on the docker network.

NOTE: this enables use of Faucet's egress pipeline feature, which is currently experimental and works only on OVS.

##### Specifying a specific VLAN to use for the mirror tunnel

`-o ovs.bridge.mirror_tunnel_vid=200`

This sets the mirror tunnel VLAN to 200. The default is 256 + the VLAN of the network.

###### Enabling DHCP

`--ipam-driver null -o ovs.bridge.dhcp=true`

docker's IP management of this network will be disabled, and instead dhcp will request and maintain a DHCP lease for each container on the network, using `udhcpc`. `udhcpc` is run outside the container's PID namespace (so the container cannot see it), but within its network namespace. The container therefore does not need any special privileges and cannot change its IP address itself.

##### Mirroring

Dovesnap provides infrastructure to do centralized mirroring - you can have dovesnap mirror the traffic for any container on a network it controls, back to a single interface (virtual or physical). This allows you to (for example) run one centralized tcpdump process that can collect all mirrored traffic.

To use physical interface `eno99` for mirroring, for example:

`$ FAUCET_PREFIX=/etc/faucet MIRROR_BRIDGE_OUT=eno99 docker-compose -f docker-compose.yml -f docker-compose-standalone.yml up -d`

If you want to mirror to a virtual interface on your host, use a veth pair. For example:

```
$ sudo ip link add odsmirrori type veth peer name odsmirroro
$ sudo ip link set dev odsmirrori up
$ sudo ip link set dev odsmirroro up
$ FAUCET_PREFIX=/etc/faucet MIRROR_BRIDGE_OUT=odsmirrori docker-compose -f docker-compose.yml -f docker-compose-standalone.yml up -d
$ sudo tcpdump -n -e -v -i odsmirroro
```

From this point, any container selected for mirroring (see below) will have traffic mirrored to tcpdump running on `mirroro`

###### Mirroring across multiple hosts

You might want to run docker and dovesnap on multiple hosts, and have all the mirrored traffic from all the hosts arrive on one port on one host.

You can do this by daisy-chaining the hosts together with dedicated physical ports in a so called "mirror river". On the hosts within the chain, specify `MIRROR_BRIDGE_IN=eth77` (where `eth77` is connected to the previous host). This will cause each host to pass the mirrored traffic along to the final host.

#### Container options

These options are supplied when starting a container.

#### ACLs

`--label="dovesnap.faucet.portacl=<aclname>"`

An ACL will be applied to the port associated with the container. The ACL must already exist in FAUCET (e.g. by adding it to `faucet.yaml`).

If a container is connected to multiple dovesnap networks, it is possible to specify different ACLs per network:

`--label="dovesnap.faucet.portacl=<networkname>:<aclname>/..."`

#### Mirroring

`--label="dovesnap.faucet.mirror=true"`

The container's traffic (both sent and received) will be mirrored to a port on the bridge (see above).

If a container is connected to multiple dovesnap networks, it is possible to specify different mirror options for each network:

`--label="dovesnap.faucet.mirror=<networkname>:<true>/..."`

#### MAC prefix

`--label="dovesnap.faucet.mac_prefix=0e:99`

The prefix of the container interface's MAC address will be replaced with the specified bytes (1 to 5 bytes may be supplied).
This can be convenient when filtering traffic with tcpdump - containers with this label will have an easily identifiable MAC address.

NOTE: where this option is used, the MAC address reported by `docker inspect` will be inaccurate.

#### Visualizing dovesnap networks

Dovesnap can generate a diagram of how containers and interfaces are connected together, with some information about running containers (e.g. MAC and IP addresses). This can be useful for troubleshooting or verifying configuration.

```
$ sudo pip3 install -r requirements.txt
$ cd bin
$ ./graph_dovesnap
```

A PNG file will be created that describes the networks dovesnap controls.


CHANGELOG.md
# v1.1.4

- c65sdn 1.0.43
- grpc 1.5.1
- OVS 3.0.1
- frpc 0.22.44
- docker 6.0.1
- grafana, prometheus, pytype

# v1.1.3

- Prometheus 2.29.0
- Grafana 9.1.7
- c65sdn 1.0.39
- frpc 0.22.42
 
# v1.1.2

- OVS 3.0.0
- golang 1.19
- grpc 1.4.9
- c65sdn 1.0.37
- logrus, grafana

# v1.1.1

- workarounds for missing OVS config/local port status.

# v1.1.0

- Updated faucetconfrpc, prometheus, grafana, faucet, gauge, openvswitch, and grpc
- Moved package to use pyproject.toml

# v1.0.10

- faucetconfrpc 0.22.35

# v1.0.9

- OVS 2.17.1
- c65sdn/faucet to 1.0.35
- golang 1.18
- libyaml 3.0
- grpc 1.46.2
- faucetconfrpc 0.22.34
- pylint, pytype, grafana, prometheus

# v1.0.8

- Update c65faucet to 1.0.29
- Update grpc to v1.45.0
- Update OVS to 2.17.0
- Update faucetconfrpc to 0.22.33
- Update prometheus, pytype, grafana

# v1.0.7

- Update c65faucet to 1.0.28
- Update containerd, image-spec (https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-23648)
- Update prometheus, pytype, grafana

# v1.0.6

- Update c65faucet to 1.0.27
- Update faucetconfrpc to 0.22.32
- Update ruamel.yaml to v0.17.21
- Update prometheus, pytype, grafana

# v1.0.5

- Update c65faucet to 1.0.23
- Update faucetconfrpc to 0.22.30
- Update containerd to 1.5.9
- Update OVS to 2.16.2
- Update grpc to 1.43.0
- Update docker to 2.3.2
- Update pylint, pytype, grafana

# v1.0.3

- Update to debian bullseye
- Update pytype, prometheus, grafana, faucet, gauge, faucetconfrpc

# v1.0.2

- Upgrade docker, ruamel.yaml, prometheus

# v1.0.1

- Fix packaging.

# v1.0.0

- Driver name has changed from ovs to dovesnap
- Add cleanup_dovesnap script
- Update faucet, gauge, grafana, pytype, pylint, faucetconfrpc, ovs

# v0.22.6

- Make Join() and CreateEndpoint() non blocking again, but retain the head start on OFPort provisioning in CreateEn
dpoint.

# v0.22.5 (2021-08-16)

- allocate OVS ports when endpoint is created, before container Join.
- allow preallocation of N FAUCET ports.
- trigger state reconcilation on webserver request for status
- upgrade pytype, prometheus, grpc

# v0.22.4 (2021-08-11)

- immediately allocate OVS port at container Join (faster networking setup)
- upgrade faucet, gauge, pytype, docker, grafana, grpc

# v0.22.3 (2021-07-28)

- reduce size of OVS and dovesnap containers (builders)
- upgrade to golang 1.16
- min TLS 1.3 version for faucetconfrpc
- pin more python dependencies (including graphviz, docker)

# v0.22.2 (2021-07-21)

- upgrade faucet, grafana, pytype
- docker as a systemd service

# v0.22.1 (2021-07-14)

- Use waitgroups to synchronize create/delete operations, including on exit.
- OVS 2.15.1
- Update faucet, gauge, pylint, grpc, grafana, docker, prometheus, faucetconfrpc

# v0.22.0 (2021-06-22)

- dovesnap plugin no longer needs privileged: true
- update prometheus, grafana, pytype.

# v0.21.0 (2021-06-17)

- restart/shutdown handling improvements, and network deletes now blocking to avoid leaked veths or other resources.
- Update pytype, docker, faucetconfrpc, prometheus, grafana, baseconv

# v0.20.0 (2021-05-04)

- Use c65sdn faucet/gauge docker images
- Upgrade pytype, docker, faucetconfrpc, prometheus, grafana
- Add routed mode

# v0.19.0 (2021-03-05)

- Use c65sdn faucet/gauge docker images
- Upgrade pytype, docker, faucetconfrpc

# v0.18.0 (2021-02-26)

- Fix mirror bridge may not be initialized correctly due to port number assignment
- Add network name specific mirror and ACL options
- Fix deadlock between Dovesnap and Docker when creating a lot of containers
- Delete all flows when creating a bridge to disable default switching
- Log all CLI arguments for debugging
- Upgrade OVS to 2.14.2
- Upgrade docker, grafana, grpc, logrus, prometheus, pytype

# v0.17.0 (2021-02-09)

- Dovesnap populates status server with DHCP addresses for containers
- graph script can generate a global view across multiple dovesnaps
- Upgrade docker, pytype

# v0.16.0 (2021-02-02)

- retry failed dump-ports-desc calls
- add VLAN output ACL
- upgrade OVS 2.14.1
- upgrade faucetconfrpc, certstrap

# v0.15.0 (2021-01-21)

- Add ovs.bridge.add_copro_ports
- Upgrade faucetconfrpc, certstrap, prometheus, grafana, grpc

# v0.14.0 (2021-01-14)

- Handle empty FAUCET config at startup
- Handle cold start of host with existing docker networks
- Handle ovs-ofctl returning an error temporarily, reconciling networks
- Upgrade prometheus, grpc, faucet, gauge, faucetconfrpc, certstrap, grafana

# v0.13.0 (2020-11-25)

- Updated grafana, faucetconfrpc, faucet-certstrap
- Fix stack test, for faucet

# v0.12.0 (2020-11-19)

- Updated prometheus, grafana, faucet, gauge, faucet-certstrap, faucetconfrpc
- Restructured tests

# v0.11.0 (2020-10-30)

- VLAN documentation
- Update faucet, gauge, certstrap, faucetconfrpc

# v0.10.0 (2020-10-22)

- Fix MANIFEST.in to include main.go
- Allow configuration of faucetconfrpc client key
- Upgrade grpc, grafana

# v0.9.0 (2020-10-21)

- Move non-error OVS debug to trace level
- Add support for setting container/NAT interface MAC prefixes.

# v0.8.0 (2020-10-09)

- add JSON endpoint to get dovesnap state
- log MAC address of containers
- Update faucet, faucetconfrpc, certstrap, grafana, grpc

# v0.7.0 (2020-09-30)

- allow ACL on NAT port
- tests check mirror flows are present
- mirror bridge should pass through packets with any VLAN tag
- fix viz crash when no ip command
- upgrade logrus, faucetconfrpc, grafana

# v0.6.0 (2020-09-22)

- fix race condition that could lead to multiple containers with same OFPort
- add NAT port mapping (-p) support via network gateway IP
- Remove OVS privilege when in userspace mode
- Move to go mod from dep ensure
- Viz script timeout/missing IP fixes
- update faucetconfrpc, OVS

# v0.5.0 (2020-09-04)

- update faucetconfrpc, certstrap
- graph_dovesnap vizualization script
- support OVS userspace mode

# v0.4.0 (2020-08-27)

- add/remove non-dovesnap managed ports to bridges to FAUCET
- ARM compatibility
- Updated grafana, faucetconfrpc

# v0.3.0 (2020-08-21)


ovs/ovs_config.go
package ovs

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/docker/docker/api/types"
	networkplugin "github.com/docker/go-plugins-helpers/network"
)

const (
	DriverName = "ovs"

	genericOption  = "com.docker.network.generic"
	internalOption = "com.docker.network.internal"
	portMapOption  = "com.docker.network.portmap"

	bindInterfaceOption    = "ovs.bridge.bind_interface"
	bridgeAddPorts         = "ovs.bridge.add_ports"
	bridgeAddCoproPorts    = "ovs.bridge.add_copro_ports"
	bridgeController       = "ovs.bridge.controller"
	bridgeDpid             = "ovs.bridge.dpid"
	bridgeLbPort           = "ovs.bridge.lbport"
	bridgeNameOption       = "ovs.bridge.name"
	dhcpOption             = "ovs.bridge.dhcp"
	mirrorTunnelVid        = "ovs.bridge.mirror_tunnel_vid"
	modeOption             = "ovs.bridge.mode"
	NATAclOption           = "ovs.bridge.nat_acl"
	mtuOption              = "ovs.bridge.mtu"
	vlanOption             = "ovs.bridge.vlan"
	userspaceOption        = "ovs.bridge.userspace"
	ovsLocalMacOption      = "ovs.bridge.ovs_local_mac"
	vlanOutAclOption       = "ovs.bridge.vlan_out_acl"
	defaultAclOption       = "ovs.bridge.default_acl"
	preAllocatePortsOption = "ovs.bridge.preallocate_ports"

	defaultLbPort           = 99
	defaultMTU              = 1500
	defaultMode             = modeFlat
	defaultRoute            = "0.0.0.0/0"
	defaultTunnelVLANOffset = 256
	defaultVLAN             = 100

	modeFlat   = "flat"
	modeNAT    = "nat"
	modeRouted = "routed"

	bridgePrefix                 = "ovsbr-"
	containerEthName             = "eth"
	netNsPath                    = "/var/run/netns"
	dhcpStatePath                = "/var/run"
	ofPortLocal       OFPortType = 4294967294
	ovsPortPrefix                = "ovs-veth0-"
	patchPrefix                  = "ovp"
	peerOvsPortPrefix            = "ethc"
	stackDpidPrefix              = "0x0E0F00"
	ovsStartupRetries            = 5
	dockerRetries                = 5
)

var (
	validModes = map[string]bool{
		modeNAT:    true,
		modeFlat:   true,
		modeRouted: true,
	}
)

type StackMirrorConfig struct {
	LbPort           OFPortType
	TunnelVid        OFVidType
	RemoteDpName     string
	RemoteMirrorPort OFPortType
}

func makeDynamicNetworkState(shortEngineId string) DynamicNetworkState {
	return DynamicNetworkState{
		ShortEngineId:    shortEngineId,
		Containers:       make(map[string]ContainerState),
		ExternalPorts:    make(map[string]ExternalPortState),
		OtherBridgePorts: make(map[string]OtherBridgePortState),
	}
}

func mustGetInternalOption(r *networkplugin.CreateNetworkRequest) bool {
	if r.Options == nil {
		return false
	}
	return r.Options[internalOption].(bool)
}

func getGenericOption(r *networkplugin.CreateNetworkRequest, optionName string) string {
	if r.Options == nil {
		return ""
	}
	optionsMap, have_options := r.Options[genericOption].(map[string]interface{})
	if !have_options {
		return ""
	}
	optionValue, have_option := optionsMap[optionName].(string)
	if !have_option {
		return ""
	}
	return optionValue
}

func getGenericUintOption(r *networkplugin.CreateNetworkRequest, optionName string, defaultOption uint) uint {
	option := getGenericOption(r, optionName)
	return uint(defaultUint(option, uint64(defaultOption)))
}

func mustGetTunnelVid(r *networkplugin.CreateNetworkRequest) uint {
	return getGenericUintOption(r, mirrorTunnelVid, mustGetBridgeVLAN(r)+defaultTunnelVLANOffset)
}

func mustGetBridgeMTU(r *networkplugin.CreateNetworkRequest) uint {
	return getGenericUintOption(r, mtuOption, defaultMTU)
}

func mustGetPreAllocatePorts(r *networkplugin.CreateNetworkRequest) uint {
	return getGenericUintOption(r, preAllocatePortsOption, 0)
}

func mustGetLbPort(r *networkplugin.CreateNetworkRequest) uint {
	return getGenericUintOption(r, bridgeLbPort, defaultLbPort)
}

func mustGetBridgeName(r *networkplugin.CreateNetworkRequest) string {
	bridgeName := bridgePrefix + truncateID(r.NetworkID)
	name := getGenericOption(r, bridgeNameOption)
	if name != "" {
		bridgeName = name
	}
	return bridgeName
}

func mustGetOvsLocalMac(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, ovsLocalMacOption)
}

func mustGetBridgeMode(r *networkplugin.CreateNetworkRequest) string {
	bridgeMode := defaultMode
	mode := getGenericOption(r, modeOption)
	if mode != "" {
		if _, isValid := validModes[mode]; !isValid {
			panic(fmt.Errorf("%s is not a valid mode", mode))
		}
		bridgeMode = mode
	}
	return bridgeMode
}

func mustGetBridgeController(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, bridgeController)
}

func mustGetBridgeDpid(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, bridgeDpid)
}

func mustGetBridgeVLAN(r *networkplugin.CreateNetworkRequest) uint {
	return getGenericUintOption(r, vlanOption, defaultVLAN)
}

func mustGetBridgeVLANOutAcl(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, vlanOutAclOption)
}

func mustGetDefaultAcl(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, defaultAclOption)
}

func mustGetBridgeAddPorts(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, bridgeAddPorts)
}

func mustGetBridgeAddCoproPorts(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, bridgeAddCoproPorts)
}

func mustGetNATAcl(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, NATAclOption)
}

func mustGetGatewayIPFromData(data []*networkplugin.IPAMData) string {
	if len(data) > 0 {
		if data[0] != nil {
			if data[0].Gateway != "" {
				return data[0].Gateway
			}
		}
	}
	return ""
}

func mustGetGatewayIP(r *networkplugin.CreateNetworkRequest) (string, string) {
	// Guess gateway IP, prefer IPv4.
	ipv6Gw := mustGetGatewayIPFromData(r.IPv6Data)
	ipv4Gw := mustGetGatewayIPFromData(r.IPv4Data)
	gatewayIP := ipv6Gw
	if ipv4Gw != "" {
		gatewayIP = ipv4Gw
	}
	if gatewayIP == "" {
		return "", ""
	}
	parts := strings.Split(gatewayIP, "/")
	if len(parts) == 2 && parts[0] != "" && parts[1] != "" {
		return parts[0], parts[1]
	}
	panic(fmt.Errorf("cannot parse gateway IP: %s", gatewayIP))
}

func mustGetBindInterface(r *networkplugin.CreateNetworkRequest) string {
	return getGenericOption(r, bindInterfaceOption)
}

func parseBool(optionVal string) bool {
	boolVal, err := strconv.ParseBool(optionVal)
	if err != nil {
		return false
	}
	return boolVal
}

func mustGetUseDHCP(r *networkplugin.CreateNetworkRequest) bool {
	return parseBool(getGenericOption(r, dhcpOption))
}

func mustGetUserspace(r *networkplugin.CreateNetworkRequest) bool {
	return parseBool(getGenericOption(r, userspaceOption))
}

func truncateID(id string) string {
	return id[:5]
}

func (d *Driver) mustGetStackBridgeLink() (string, OFPortType) {
	return d.stackMirrorInterface[0], OFPortType(defaultUint(d.stackMirrorInterface[1], 0))
}

func (d *Driver) getStackMirrorConfig(r *networkplugin.CreateNetworkRequest) StackMirrorConfig {
	lbPort := mustGetLbPort(r)
	var tunnelVid uint = 0
	remoteDpName := ""
	var mirrorPort OFPortType = 0

	if usingStackMirroring(d) {
		tunnelVid = mustGetTunnelVid(r)
		remoteDpName, mirrorPort = d.mustGetStackBridgeLink()
	}

	return StackMirrorConfig{
		LbPort:           OFPortType(lbPort),
		TunnelVid:        OFVidType(tunnelVid),
		RemoteDpName:     remoteDpName,
		RemoteMirrorPort: OFPortType(mirrorPort),
	}
}

func (d *Driver) mustGetStackDPName() string {
	return "dovesnap" + d.shortEngineId
}

func (d *Driver) mustGetStackDP() (string, string) {
	dpid := stackDpidPrefix + d.shortEngineId
	return dpid, d.mustGetStackDPName()
}

func (d *Driver) mustGetLoopbackBrName() string {
	return "lb" + d.shortEngineId
}

func (d *Driver) mustGetMirrorBrName() string {
	return "mirror" + d.shortEngineId
}

func (d *Driver) mustGetStackingInterface(stackingInterface string) (string, OFPortType, string) {
	stackSlice := strings.Split(stackingInterface, ":")
	remoteDP := stackSlice[0]
	remotePort, err := ParseUint32(stackSlice[1])
	if err != nil {
		panic(fmt.Errorf("unable to convert remote port to an unsigned integer because: [ %s ]", err))
	}
	localInterface := stackSlice[2]
	if err != nil {
		panic(fmt.Errorf("unable to convert local port to an unsigned integer because: [ %s ]", err))
	}
	return remoteDP, OFPortType(remotePort), localInterface
}

func (d *Driver) mustGetStackBridgeConfig() (string, string, uint64, string) {
	dpid, dpName := d.mustGetStackDP()
	hostname, err := os.Hostname()
	if err != nil {
		panic(err)
	}
	uintDpid := mustGetUintFromHexStr(dpid)
	return hostname, dpid, uintDpid, dpName
}

func mustGetBridgeNameFromResource(r *types.NetworkResource) string {
	return bridgePrefix + truncateID(r.ID)
}

func getStrOptionFromResource(r *types.NetworkResource, optionName string, defaultOptionValue string) string {
	if r.Options == nil {
		return defaultOptionValue
	}
	optionValue, have_option := r.Options[optionName]
	if !have_option {
		return defaultOptionValue
	}
	return optionValue
}

func getUintOptionFromResource(r *types.NetworkResource, optionName string, defaultOptionValue uint) uint {
	optionStrValue := getStrOptionFromResource(r, optionName, "")
	return uint(defaultUint(optionStrValue, uint64(defaultOptionValue)))
}

func mustGetBridgeDpidFromResource(r *types.NetworkResource) (string, uint64) {
	dpid := getStrOptionFromResource(r, bridgeDpid, "")
	uintDpid := mustGetUintFromHexStr(dpid)
	return dpid, uintDpid
}

func getGatewayFromResource(r *types.NetworkResource) (string, string) {
	if len(r.IPAM.Config) > 0 {
		config := r.IPAM.Config[0]
		subnetIP := config.Subnet
		parts := strings.Split(subnetIP, "/")
		if len(parts) == 2 && parts[0] != "" && parts[1] != "" {
			return config.Gateway, parts[1]
		}
	}
	return "", ""
}

func getStrForNetwork(networkStr string, networkName string) string {
	networkStrs := ""
	networksStrsList := strings.Split(networkStr, "/")
	for _, networksStrsPairStr := range networksStrsList {
		networkStrsPair := strings.Split(networksStrsPairStr, ":")
		if len(networkStrsPair) == 1 {
			networkStrs = networkStrsPair[0]
		} else if len(networkStrsPair) == 2 {
			if networkName == networkStrsPair[0] {
				networkStrs = networkStrsPair[1]
			}
		}
	}
	return networkStrs
}

func getNetworkStateFromResource(r *types.NetworkResource, shortEngineId string) (NetworkState, error) {
	var err error = nil
	ns := NetworkState{}
	defer func() {
		if rerr := recover(); rerr != nil {
			err = fmt.Errorf("missing bridge info: %v", rerr)
		}
	}()
	dpid, uintDpid := mustGetBridgeDpidFromResource(r)
	gateway, mask := getGatewayFromResource(r)
	ns = NetworkState{
		NetworkName:          r.Name,
		BridgeName:           mustGetBridgeNameFromResource(r),
		BridgeDpid:           dpid,
		BridgeDpidUint:       uintDpid,
		BridgeVLAN:           getUintOptionFromResource(r, vlanOption, defaultVLAN),
		MTU:                  getUintOptionFromResource(r, mtuOption, defaultMTU),
		PreAllocatePorts:     getUintOptionFromResource(r, preAllocatePortsOption, 0),
		Mode:                 getStrOptionFromResource(r, modeOption, defaultMode),
		FlatBindInterface:    getStrOptionFromResource(r, bindInterfaceOption, ""),
		AddPorts:             getStrOptionFromResource(r, bridgeAddPorts, ""),
		AddCoproPorts:        getStrOptionFromResource(r, bridgeAddCoproPorts, ""),
		UseDHCP:              parseBool(getStrOptionFromResource(r, dhcpOption, "")),
		Userspace:            parseBool(getStrOptionFromResource(r, userspaceOption, "")),
		Gateway:              gateway,
		GatewayMask:          mask,
		NATAcl:               getStrOptionFromResource(r, NATAclOption, ""),
		VLANOutAcl:           getStrOptionFromResource(r, vlanOutAclOption, ""),
		DefaultAcl:           getStrOptionFromResource(r, defaultAclOption, ""),
		OvsLocalMac:          getStrOptionFromResource(r, ovsLocalMacOption, ""),
		Controller:           getStrOptionFromResource(r, bridgeController, ""),
		DynamicNetworkStates: makeDynamicNetworkState(shortEngineId),
	}
	return ns, err
}

func (d *Driver) getStackMirrorConfigFromResource(r *types.NetworkResource) StackMirrorConfig {
	lbPort := getUintOptionFromResource(r, bridgeLbPort, defaultLbPort)
	var tunnelVid uint = 0
	remoteDpName := ""
	var mirrorPort OFPortType = 0

	if usingStackMirroring(d) {
		vlan := getUintOptionFromResource(r, vlanOption, defaultVLAN)
		tunnelVid = getUintOptionFromResource(r, mirrorTunnelVid, vlan+defaultTunnelVLANOffset)
		remoteDpName, mirrorPort = d.mustGetStackBridgeLink()
	}

	return StackMirrorConfig{
		LbPort:           OFPortType(lbPort),
		TunnelVid:        OFVidType(tunnelVid),
		RemoteDpName:     remoteDpName,
		RemoteMirrorPort: OFPortType(mirrorPort),
	}
}
